File: 0141-linked-list-cycle/0141-linked-list-cycle.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun hasCycle(head: ListNode?): Boolean {
        val set = HashSet<ListNode>()
        var h = head
        while (h != null) {
            if (set.contains(h)) return true
            set.add(h)
            h = h.next
        }
        return false
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    if (!head || !head.next) return false;
    let slow = head;
    let fast = head.next;
    while (slow !== fast) {
        if (!slow || !fast || !slow.next || !fast.next) return false;
        slow = slow.next;
        fast = fast.next.next;
    }
    return true;
};